flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/conversion/DataStructureConverters.java

```diff
@@ -0,0 +1,224 @@
+		putConverter(LogicalTypeRoot.ARRAY, long[].class, singleton(ArrayLongArrayConverter::new));
+		putConverter(LogicalTypeRoot.ARRAY, float[].class, singleton(ArrayFloatArrayConverter::new));
+		putConverter(LogicalTypeRoot.ARRAY, double[].class, singleton(ArrayDoubleArrayConverter::new));
+		putConverter(LogicalTypeRoot.MAP, Map.class, MapMapConverter::createForMapType);
+		putConverter(LogicalTypeRoot.MAP, MapData.class, identity());
+		putConverter(LogicalTypeRoot.MULTISET, Map.class, MapMapConverter::createForMultisetType);
+		putConverter(LogicalTypeRoot.MULTISET, MapData.class, identity());
+		putConverter(LogicalTypeRoot.ROW, Row.class, RowRowConverter::create);
+		putConverter(LogicalTypeRoot.ROW, RowData.class, identity());
+		putConverter(LogicalTypeRoot.STRUCTURED_TYPE, Row.class, RowRowConverter::create);
+		putConverter(LogicalTypeRoot.STRUCTURED_TYPE, RowData.class, identity());
+		putConverter(LogicalTypeRoot.RAW, byte[].class, RawByteArrayConverter::create);
+		putConverter(LogicalTypeRoot.RAW, RawValueData.class, identity());
+	}
+
+	/**
+	 * Returns a converter for the given {@link DataType}.
+	 */
+	@SuppressWarnings("unchecked")
+	public static DataStructureConverter<Object, Object> getConverter(DataType dataType) {
+		// cast to Object for ease of use
+		return (DataStructureConverter<Object, Object>) getConverterInternal(dataType);
+	}
+
+	private static DataStructureConverter<?, ?> getConverterInternal(DataType dataType) {
+		final LogicalType logicalType = dataType.getLogicalType();
+		final DataStructureConverterFactory factory = converters.get(
+			new ConverterIdentifier<>(
+				logicalType.getTypeRoot(),
+				dataType.getConversionClass()));
+		if (factory != null) {
+			return factory.createConverter(dataType);
+		}
+		// special cases
+		switch (logicalType.getTypeRoot()) {
+			case ARRAY:
+				return ArrayObjectArrayConverter.create(dataType);
+			case DISTINCT_TYPE:
+				return getConverterInternal(dataType.getChildren().get(0));
+			case STRUCTURED_TYPE:
+				return StructuredObjectConverter.create(dataType);
+			case RAW:
+				return RawObjectConverter.create(dataType);
+			default:
+				throw new TableException("Could not find converter for data type: " + dataType);
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	// Helper methods
+	// --------------------------------------------------------------------------------------------
+
+	private static <E> void putConverter(
+			LogicalTypeRoot root,
+			Class<E> conversionClass,
+			DataStructureConverterFactory factory) {
+		converters.put(new ConverterIdentifier<>(root, conversionClass), factory);
+	}
+
+	private static DataStructureConverterFactory identity() {
+		return singleton(IdentityConverter::new);
+	}
+
+	private static DataStructureConverterFactory singleton(Supplier<DataStructureConverter<?, ?>> supplier) {
```

Comment:
Why is it called singleton? If I am not mistaken it will create a new converter each time the `getConverterInternal` is called.